common/lockless_queue.c

```diff
@@ -0,0 +1,143 @@
+void oe_lockless_queue_node_init(oe_lockless_queue_node* p_node)
+{
+    p_node->p_link = NULL;
+} /* init_oe_lockless_queue_node */
+
+/* functions for oe_lockless_queue */
+/*---------------------------------------------------------------------------*/
+void oe_lockless_queue_init(oe_lockless_queue* p_queue)
+{
+#ifdef _MSC_VER
+    _InterlockedExchangePointer(&(p_queue->p_tail), NULL);
+    _InterlockedExchangePointer(&(p_queue->p_head), NULL);
+#elif defined __GNUC__
+    __atomic_store_n(&(p_queue->p_tail), NULL, __ATOMIC_RELAXED);
+    __atomic_store_n(&(p_queue->p_head), NULL, __ATOMIC_RELAXED);
+#endif /* __GNUC__ */
+} /* init_oe_lockless_queue */
+
+void oe_lockless_queue_push_back(
+    oe_lockless_queue* p_queue,
+    oe_lockless_queue_node* p_node)
+{
+    oe_lockless_queue_node* p_expected = NULL;
+#ifdef _MSC_VER
+    oe_lockless_queue_node* p_actual = NULL;
+#endif /* _MSC_VER */
+    do
+    {
+#ifdef _MSC_VER
+        p_expected = p_actual;
+        p_node->p_link = p_expected;
+        p_actual = (oe_lockless_queue_node*)_InterlockedCompareExchangePointer(
+            &(p_queue->p_tail), p_node, p_expected);
+    } while (p_actual != p_expected);
+#elif defined __GNUC__
+        p_node->p_link = p_expected;
+    } while (!__atomic_compare_exchange_n(
+        &(p_queue->p_tail),
+        &p_expected,
+        p_node,
+        1,
+        __ATOMIC_ACQ_REL,
+        __ATOMIC_ACQUIRE));
+#endif /* __GNUC__ */
+} /* oe_lockless_queue_push */
+
+oe_lockless_queue_node* oe_lockless_queue_pop_front(oe_lockless_queue* p_queue)
+{
+    /* try to take a node from the head */
+    oe_lockless_queue_node* popped_node = NULL;
+#ifdef _MSC_VER
+    popped_node = (oe_lockless_queue_node*)_InterlockedCompareExchangePointer(
+        &(p_queue->p_head), NULL, NULL);
+#elif defined __GNUC__
+    popped_node = __atomic_load_n(&(p_queue->p_head), __ATOMIC_ACQUIRE);
+#endif /* __GNUC__ */
+
+    if (NULL != popped_node)
+    {
+        /* there was a node at the head
+         * pop the node from the head and replace it with the node that it
+         *     points to
+         * remove the reference from the popped node to the next node */
+        oe_lockless_queue_node* next_node = popped_node->p_link;
+        popped_node->p_link = NULL;
+#ifdef _MSC_VER
+        _InterlockedExchangePointer(&(p_queue->p_head), next_node);
+#elif defined __GNUC__
+        __atomic_store_n(&(p_queue->p_head), next_node, __ATOMIC_RELEASE);
+#endif /* __GNUC__ */
+    }
+    else
+    {
+        /* there wasn't a node at the head
+         * so refill the head with the nodes from the tail */
+#ifdef _MSC_VER
+        popped_node =
+            (oe_lockless_queue_node*)_InterlockedCompareExchangePointer(
+                &(p_queue->p_tail), NULL, NULL);
+#elif defined __GNUC__
+        popped_node = __atomic_load_n(&(p_queue->p_tail), __ATOMIC_RELAXED);
+#endif /* __GNUC__ */
+
+        if (NULL != popped_node)
```

Comment:
If the head-stack is empty, we are always going to fill the head-stack with reversed tail-stack.
In that case, it seems like this entire logic can be simplified by atomically swapping p_tail and NULL:

```c
popped_node = NULL;
oe_lockless_queue_node* null_node = NULL;
__atomic_exchange(&(p_queue->p_tail), &null_node, &popped_node, __ATOMIC_ACQUIRE);
```
// Now that the list has been detached from tail, reverse it if not empty and install as head-stack.


Am I missing something?

Reply:
I noticed that too and intended to change that logic.  I must have gotten distracted.  The old gnu intrinsic libs didn't have a simple exchange (only compare exchange).  I have made this change which is now running in sandbox.